internal/controller/stas/containerimagescan_controller.go

```diff
@@ -49,12 +50,27 @@ func (r *ContainerImageScanReconciler) Reconcile(ctx context.Context, req ctrl.R
 			return ctrl.Result{}, staserrors.Ignore(err, apierrors.IsNotFound)
 		}
 
-		timeUntilNextScan := r.TimeUntilNextScan(cis)
-		if timeUntilNextScan > 0 {
-			return ctrl.Result{RequeueAfter: timeUntilNextScan}, nil
+		var timeUntilNextScan time.Duration
+
+		switch {
+		case cis.Status.ObservedGeneration != cis.Generation:
+			// Spec has changed; must scan
+		case cis.Status.LastScanTime.IsZero():
+			// New resource; must scan
+		default:
+			d := time.Until(cis.Status.LastScanTime.Add(r.ScanInterval))
+			if d > 0 {
+				timeUntilNextScan = d
+			}
+		}
+
+		var err error
+		if timeUntilNextScan == 0 {
+			err = r.reconcile(ctx, cis)
+			timeUntilNextScan = r.ScanInterval
```

Comment:
Is the above logic same as the one below?

```suggestion
		var timeUntilNextScan time.Duration
		var err error

		d := time.Until(cis.Status.LastScanTime.Add(r.ScanInterval))
		if d > 0 {
			timeUntilNextScan = d
		} else {
			err = r.reconcile(ctx, cis)
			timeUntilNextScan = r.ScanInterval
			}
```

Reply:
No, I don't think so:

1. Does not handle generation changed.
2. Will panic if LastScanTime is nil.

Reply:
We are mostly interested in LastScanTime, but we need to override this behavior when spec is changed.

Reply:
We could use a if-else instead, if that helps?

Reply:
Yes. Also updated the suggestion to use 2 if's.

Reply:
But that is not enough. We must handle the generation changed.

Reply:
Using predicates now, and I think the logic became clearer. PTAL!